### PR TITLE
Cache Error Message Detail

### PR DESCRIPTION
**Pull request by manadart**
## Description of change

This patch modifies errors emitted by cached units for unset charm URL. Message now indicate the operation that was attempted.

## QA steps

No functional change. Unit tests verify new error text.

## Documentation changes

None.

## Bug reference

N/A

### CODE REVIEW

**Comment by manadart** [comment]
!!build!!

**Comment by manadart** [comment]
$$merge$$

**Comment by manadart** [comment]
[Charm-not-set Messages Specific to Operation Attempted](https://trello.com/c/2m8U1R4G/1622-charm-not-set-messages-specific-to-operation-attempted)

File: core/cache/unit.go [hunk 1]
```diff
@@ -74,7 +74,7 @@ func (u *Unit) Ports() []network.Port {
 // taking into account whether it is tracking a model branch.
 func (u *Unit) ConfigSettings() (charm.Settings, error) {
 	if u.details.CharmURL == "" {
-		return nil, errors.New("unit charm not set")
+		return nil, errors.New("unit charm must be set before retrieving config")
```
Comment (jameinel) [inline core/cache/unit.go]: Is 'unit charm' meaningful to people? It sounds a bit odd. vs "unit's charm URL" or something along those lines.

Comment (manadart) [reply core/cache/unit.go]: Done as suggested.

**Comment by SimonRichardson** [comment]
!!build!!

**Comment by SimonRichardson** [comment]
!!build!!

**Comment by SimonRichardson** [comment]
$$merge$$

**Comment by SimonRichardson** [comment]
!!build!!

**Comment by SimonRichardson** [comment]
$$merge$$

**Comment by jameinel** [comment]
This feels like a genuine failure:
[LOG] 0:01.748 DEBUG juju.apiserver -> [2] unit-wordpress-0 800.43µs {"request-id":3,"response":"'body redacted'"} Uniter[""].WatchConfigSettingsHash
unit_test.go:536:
    c.Assert(err, gc.ErrorMatches, "unit's charm URL must be set before watching config")
... error string = "unit charm not set"
... regex string = "unit's charm URL must be set before watching config"

**Comment by manadart** [comment]
$$merge$$

**Pull request merged**